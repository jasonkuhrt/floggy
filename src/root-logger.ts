import * as Logger from './logger.js'
import * as Settings from './settings.js'

export type SettingsManager = Settings.Data & {
  (newSettings: Settings.Input): RootLogger
}

// TODO jsDoc for each option
export type Options = Settings.Input

export type RootLogger = Logger.Logger & {
  settings: SettingsManager
}

export type State = {
  settings: SettingsManager
}

/**
 * Create a root logger.
 */
export const create = (opts?: Options): RootLogger => {
  const settings = Settings.create(opts)
  const settingsManager = ((newSettings) => {
    settings(newSettings)
    Object.assign(settingsManager, settings)
    return logger
  }) as SettingsManager
  Object.assign(settingsManager, settings)
  const loggerLink = Logger.create({ settings: settingsManager }, null, undefined)
  const logger = loggerLink.logger as RootLogger
  logger.settings = settingsManager
  return logger
}
