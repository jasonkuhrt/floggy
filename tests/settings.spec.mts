import * as Logger from '../src/index.mjs'
import * as RootLogger from '../src/root-logger.mjs'
import {
  createMemoryOutput,
  MemoryOutput,
  mockConsoleLog,
  resetBeforeEachTest,
  unmockConsoleLog
} from './__helpers.mjs'

let l: Logger.RootLogger
let output: MemoryOutput

resetBeforeEachTest(process, 'env')

beforeEach(() => {
  process.env.LOG_PRETTY = 'false'
  output = createMemoryOutput()
  l = RootLogger.create({ output, pretty: { timeDiff: false } })
  process.stdout.columns = 200
})

describe('pretty', () => {
  describe('.enabled', () => {
    it('can be disabled', () => {
      expect(RootLogger.create({ pretty: { enabled: false } }).settings.pretty.enabled).toEqual(false)
    })
    it('persists across peer field changes', () => {
      const l = RootLogger.create({ pretty: { enabled: false } })
      l.settings({ pretty: { color: false } })
      expect(l.settings.pretty).toEqual({
        enabled: false,
        color: false,
        levelLabel: false,
        timeDiff: true
      })
    })
    // it('controls if logs are rendered pretty or as JSON', () => {
    //   if (!process.version.match(/^v12/)) return
    //   output.captureConsoleLog()
    //   log.info('foo')
    //   Logger.demo(log)
    //   expect(output.memoryOrRaw).toMatchSnapshot()
    // })
    describe('precedence', () => {
      it('considers instnace time config first', () => {
        process.stdout.isTTY = false
        process.env.LOG_PRETTY = 'false'
        const l = RootLogger.create({ pretty: false })
        l.settings({ pretty: true })
        expect(l.settings.pretty.enabled).toEqual(true)
      })
      it('then considers construction time config', () => {
        process.stdout.isTTY = false
        process.env.LOG_PRETTY = 'false'
        const l = RootLogger.create({ pretty: true })
        expect(l.settings.pretty.enabled).toEqual(true)
      })
      it('then considers LOG_PRETTY env var true|false (case insensitive)', () => {
        process.stdout.isTTY = false
        process.env.LOG_PRETTY = 'tRuE'
        const l = RootLogger.create()
        expect(l.settings.pretty.enabled).toEqual(true)
      })
      it('then defaults to process.stdout.isTTY', () => {
        delete process.env.LOG_PRETTY // pre-test logic forces false otherwise
        process.stdout.isTTY = true
        const l = RootLogger.create()
        expect(l.settings.pretty.enabled).toEqual(true)
      })
    })
  })

  describe('.color', () => {
    it('can be enabled', () => {
      l.settings({ pretty: { enabled: true, color: true } })
      l.info('foo', { qux: true })
      expect(l.settings.pretty.color).toEqual(true)
    })
    it('can be disabled', () => {
      expect(RootLogger.create({ pretty: { enabled: false, color: false } }).settings.pretty.color).toEqual(
        false
      )
    })
    it('is true by default', () => {
      expect(RootLogger.create({ pretty: { enabled: true } }).settings.pretty).toEqual({
        enabled: true,
        color: true,
        levelLabel: false,
        timeDiff: true
      })
      expect(RootLogger.create({ pretty: { enabled: false } }).settings.pretty).toEqual({
        enabled: false,
        color: true,
        levelLabel: false,
        timeDiff: true
      })
    })
    it('persists across peer field changes', () => {
      const l = RootLogger.create({
        pretty: { enabled: false, color: false }
      })
      l.settings({ pretty: { enabled: true } })
      expect(l.settings.pretty).toEqual({
        enabled: true,
        color: false,
        levelLabel: false,
        timeDiff: true
      })
    })
  })
  describe('.levelLabel', () => {
    it('is false by default', () => {
      expect(RootLogger.create().settings.pretty.levelLabel).toEqual(false)
    })
    it('can be enabled', () => {
      expect(RootLogger.create({ pretty: { levelLabel: true } }).settings.pretty.levelLabel).toEqual(true)
      expect(
        RootLogger.create().settings({ pretty: { levelLabel: true } }).settings.pretty.levelLabel
      ).toEqual(true)
    })
    it('persists across peer field changes', () => {
      const l = RootLogger.create({ pretty: { levelLabel: true } })
      l.settings({ pretty: false })
      expect(l.settings.pretty.levelLabel).toBe(true)
    })
    it('controls if label is spelt out or not', () => {
      l.settings({
        pretty: { enabled: true, levelLabel: true, color: false }
      })
      expect(l.settings.pretty.levelLabel).toEqual(true)
    })
  })
  describe('.timeDiff', () => {
    it('is true by default', () => {
      expect(RootLogger.create().settings.pretty.timeDiff).toEqual(true)
    })
    it('can be disabled', () => {
      const l1 = RootLogger.create({ pretty: { timeDiff: false } })
      expect(l1.settings.pretty.timeDiff).toEqual(false)
      const l2 = RootLogger.create()
      l2.settings({ pretty: { timeDiff: false } })
      expect(l2.settings.pretty.levelLabel).toEqual(false)
    })
    it('persists across peer field changes', () => {
      const l = RootLogger.create({ pretty: { timeDiff: false } })
      l.settings({ pretty: false })
      expect(l.settings.pretty.timeDiff).toBe(false)
    })
    it('can be enabled', () => {
      l.settings({
        pretty: { enabled: true, color: false, timeDiff: true }
      })
      expect(l.settings.pretty.timeDiff).toBe(true)
    })
    // todo these tests as unit level to some pure logic functions would be
    // easy... e.g. prettifier.spec.ts ... But then we run the risk of sliding
    // toward testing internals too much :\
    it.todo('renders as secodns if >= 10s')
    it.todo('renders as minutes if >= 100s')
    it.todo('renders as hours if >= 60m')
    it.todo('renders as days if >= 24h')
  })

  describe('shorthands', () => {
    it('true means enabled true', () => {
      expect(RootLogger.create({ pretty: true }).settings.pretty).toEqual({
        enabled: true,
        color: true,
        levelLabel: false,
        timeDiff: true
      })
    })

    it('false means enabled false', () => {
      expect(RootLogger.create({ pretty: false }).settings.pretty).toEqual({
        enabled: false,
        color: true,
        levelLabel: false,
        timeDiff: true
      })
    })
  })
})

describe('filter', () => {
  describe('precedence', () => {
    it('considers instance time config first', () => {
      process.env.LOG_FILTER = 'from_env_var'
      const l = RootLogger.create({ filter: { pattern: '*@fatal' } })
      l.settings({ filter: { pattern: 'foo' } })
      expect(l.settings.filter.patterns).toMatchSnapshot()
    })

    it('then considers construction time config', () => {
      process.env.LOG_FILTER = 'from_env_var'
      const l = RootLogger.create({ filter: { pattern: '*@fatal' } })
      expect(l.settings.filter).toMatchSnapshot()
    })

    it('then considers LOG_FILTER env var', () => {
      process.env.LOG_FILTER = 'from_env_var'
      const l = RootLogger.create()
      expect(l.settings.filter).toMatchSnapshot()
    })

    it('then defaults to "*', () => {
      const l = RootLogger.create()
      expect(l.settings.filter).toMatchSnapshot()
    })
  })

  it('can be passed a pattern directly', () => {
    const l = RootLogger.create({ filter: '*@fatal' })
    expect(l.settings.filter.patterns[0]?.level.value).toBe('fatal')
    l.settings({ filter: 'foo' })
    expect(l.settings.filter.patterns[0]?.level.value).toBe('debug')
    l.settings({ filter: { level: 'warn' } })
    expect(l.settings.filter.patterns[0]?.level.value).toBe('warn')
    l.settings({ filter: 'bar' })
    expect(l.settings.filter.patterns[0]?.level.value).toBe('warn')
  })

  it('LOG_FILTER envar config when invalid triggers readable log warning', () => {
    const calls = mockConsoleLog()
    process.env.LOG_FILTER = '**'
    RootLogger.create()
    expect(calls).toMatchSnapshot()
    unmockConsoleLog()
  })
})

describe('level', () => {
  describe('precedence', () => {
    it('considers instance time config first', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'fatal'
      const l = RootLogger.create({ filter: { level: 'fatal' } })
      l.settings({ filter: { level: 'trace' } })
      expect(l.settings.filter.patterns[0]?.level.value).toEqual('trace')
    })

    it('then considers construction time config', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'fatal'
      const l = RootLogger.create({ filter: { level: 'trace' } })
      expect(l.settings.filter.patterns[0]?.level.value).toEqual('trace')
    })

    it('then considers LOG_LEVEL env var', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'trace'
      const l = RootLogger.create()
      expect(l.settings.filter.patterns[0]?.level.value).toEqual('trace')
    })

    it('then considers NODE_ENV=production', () => {
      process.env.NODE_ENV = 'production'
      const l = RootLogger.create()
      expect(l.settings.filter.patterns[0]?.level.value).toEqual('info')
    })

    it('then defaults to debug', () => {
      const l = RootLogger.create()
      expect(l.settings.filter.patterns[0]?.level.value).toEqual('debug')
    })
  })

  it('logs below set level are not output', () => {
    l.settings({ filter: { level: 'warn' } }).info('foo')
    expect(output.memory).toEqual([])
  })

  it('LOG_LEVEL env var config is treated case insensitive', () => {
    process.env.NODE_ENV = 'production'
    process.env.LOG_LEVEL = 'TRACE'
    const l = RootLogger.create()
    expect(l.settings.filter.patterns[0]?.level.value).toEqual('trace')
  })

  it('LOG_LEVEL env var config when invalid triggers thrown readable error', () => {
    process.env.LOG_LEVEL = 'ttrace'
    expect(() => RootLogger.create()).toThrowErrorMatchingInlineSnapshot(
      `"Could not parse environment variable LOG_LEVEL into LogLevel. The environment variable was: ttrace. A valid environment variable must be like: fatal, error, warn, info, debug, trace"`
    )
  })
})

describe('data', () => {
  it('all defaults to false if process.NODE_ENV is not production', () => {
    expect(RootLogger.create().settings.data).toMatchInlineSnapshot(`
      {
        "hostname": false,
        "pid": false,
        "time": false,
      }
    `)
    process.env.NODE_ENV = 'not production'
    expect(RootLogger.create().settings.data).toMatchInlineSnapshot(`
      {
        "hostname": false,
        "pid": false,
        "time": false,
      }
    `)

    process.env.NODE_ENV = 'production'
    expect(RootLogger.create().settings.data).toMatchInlineSnapshot(`
      {
        "hostname": true,
        "pid": true,
        "time": true,
      }
    `)
  })

  it('merges initial setting with defaults', () => {
    expect(RootLogger.create({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'not production'
    expect(RootLogger.create({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'production'
    expect(RootLogger.create({ data: { time: false } }).settings.data).toMatchInlineSnapshot(`
      {
        "hostname": true,
        "pid": true,
        "time": false,
      }
    `)
  })

  it('merges incremental changes with previous state', () => {
    expect(RootLogger.create().settings({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'not production'
    expect(RootLogger.create().settings({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'production'
    expect(RootLogger.create().settings({ data: { time: false } }).settings.data).toMatchInlineSnapshot(`
      {
        "hostname": true,
        "pid": true,
        "time": false,
      }
    `)
  })
})
