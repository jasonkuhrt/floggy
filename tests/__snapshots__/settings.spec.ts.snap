// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter LOG_FILTER envar config when invalid triggers readable log warning 1`] = `
[
  [
    "Your log filter's pattern found in environment variable LOG_FILTER. was invalid: "**"

LOG FILTERING SYNTAX MANUAL  ⟁
▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬

Grammar

    [!][<root>](*|<path>[:*])[@(*|(<levelNum>|<levelLabel>)[+-])] [,<pattern>]

    <root>       = .
    <path>       = /^[A-z_]+[A-z_0-9]*$/ [:<path>]
    <levelNum>   = 1     | 2     | 3    | 4    | 5     | 6
    <levelLabel> = trace | debug | info | warn | error | fatal

Examples

    All logs at...

    Path
    app         app path at default level
    app:router  app:router path at default level 

    List
    app,nexus   app and nexus paths at default level

    Path Wildcard
    *           any path at default level 
    app:*       app path plus descendants at defualt level
    app::*      app path descendants at defualt level

    Negation
    !app      any path at any level _except_ those at app path at default level 
    !*        no path (meaning, nothing will be logged) 

    Removal
    *,!app      any path at default level _except_ logs at app path at default level 
    *,!*@2-     any path _except_ those at debug level or lower 
    app,!app@4  app path at defualt level _except_ those at warn level 

    Levels
    *@info      all paths at info level
    *@error-    all paths at error level or lower
    *@debug+    all paths at debug level or higher
    *@3         all paths at info level
    *@4-        all paths at error level or lower
    *@2+        all paths at debug level or higher
    app:*@2-    app path plus descendants at debug level or lower 
    app::*@2+   app path descendants at debug level or higher

    Level Wildcard
    app@*       app path at all levels
    *@*         all paths at all levels

    Explicit Root
    .           root path at defualt level
    .@info      root path at info level
    .:app       Same as "app"
    .:*         Same as "*"
  ",
  ],
]
`;

exports[`filter precedence considers instance time config first 1`] = `
[
  {
    "level": {
      "comp": "gte",
      "value": "debug",
    },
    "negate": false,
    "originalInput": "foo",
    "path": {
      "descendants": false,
      "value": ".:foo",
    },
  },
]
`;

exports[`filter precedence then considers LOG_FILTER env var 1`] = `
{
  "defaults": {
    "level": {
      "comp": "gte",
      "value": "debug",
    },
  },
  "originalInput": undefined,
  "patterns": [
    {
      "level": {
        "comp": "gte",
        "value": "debug",
      },
      "negate": false,
      "originalInput": "from_env_var",
      "path": {
        "descendants": false,
        "value": ".:from_env_var",
      },
    },
  ],
}
`;

exports[`filter precedence then considers construction time config 1`] = `
{
  "defaults": {
    "level": {
      "comp": "gte",
      "value": "debug",
    },
  },
  "originalInput": "*@fatal",
  "patterns": [
    {
      "level": {
        "comp": "eq",
        "value": "fatal",
      },
      "negate": false,
      "originalInput": "*@fatal",
      "path": {
        "descendants": {
          "includeParent": true,
        },
        "value": ".",
      },
    },
  ],
}
`;

exports[`filter precedence then defaults to "* 1`] = `
{
  "defaults": {
    "level": {
      "comp": "gte",
      "value": "debug",
    },
  },
  "originalInput": "*",
  "patterns": [
    {
      "level": {
        "comp": "gte",
        "value": "debug",
      },
      "negate": false,
      "originalInput": "*",
      "path": {
        "descendants": {
          "includeParent": true,
        },
        "value": ".",
      },
    },
  ],
}
`;
